### ISSUE TITLE
Timeout for feed iteration

### ISSUE THREAD

**Issue by ArloL** [w3stling/rssreader#163]
Thanks very much for maintaining this bit of software. It is not only a pleasure to work with but also very reliable. It executes nicely approximately every 25 minutes on my machine 👌

Sometimes though it freezes while iterating on the stream. I have attached a thread dump. My preliminary analysis is that the HttpClient timeouts are not applied because the request headers were received; also the .get timeout on the readAsync does not apply because it has already received a Stream that it is then iterating upon. I don't have a smart idea right now on how to solve this since java's HttpClient does not allow setting e.g. the socket timeout. At least not that I found.

Maybe there is another way?

```
java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@22.0.2/Native Method)
        - parking to wait for  <0x000000070202bed0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.park(java.base@22.0.2/LockSupport.java:371)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionNode.block(java.base@22.0.2/AbstractQueuedSynchronizer.java:519)
        at java.util.concurrent.ForkJoinPool.unmanagedBlock(java.base@22.0.2/ForkJoinPool.java:4011)
        at java.util.concurrent.ForkJoinPool.managedBlock(java.base@22.0.2/ForkJoinPool.java:3959)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@22.0.2/AbstractQueuedSynchronizer.java:1707)
        at java.util.concurrent.ArrayBlockingQueue.take(java.base@22.0.2/ArrayBlockingQueue.java:420)
        at jdk.internal.net.http.ResponseSubscribers$HttpResponseInputStream.current(java.net.http@22.0.2/ResponseSubscribers.java:458)
        at jdk.internal.net.http.ResponseSubscribers$HttpResponseInputStream.read(java.net.http@22.0.2/ResponseSubscribers.java:508)
        at java.util.zip.InflaterInputStream.fill(java.base@22.0.2/InflaterInputStream.java:264)
        at java.util.zip.InflaterInputStream.read(java.base@22.0.2/InflaterInputStream.java:175)
        at java.util.zip.GZIPInputStream.read(java.base@22.0.2/GZIPInputStream.java:128)
        at java.io.BufferedInputStream.fill(java.base@22.0.2/BufferedInputStream.java:291)
        at java.io.BufferedInputStream.read1(java.base@22.0.2/BufferedInputStream.java:347)
        at java.io.BufferedInputStream.implRead(java.base@22.0.2/BufferedInputStream.java:420)
        at java.io.BufferedInputStream.read(java.base@22.0.2/BufferedInputStream.java:399)
        at com.sun.org.apache.xerces.internal.impl.XMLEntityManager$RewindableInputStream.read(java.xml@22.0.2/XMLEntityManager.java:3124)
        at com.sun.org.apache.xerces.internal.impl.io.UTF8Reader.read(java.xml@22.0.2/UTF8Reader.java:303)
        at com.sun.org.apache.xerces.internal.impl.XMLEntityScanner.load(java.xml@22.0.2/XMLEntityScanner.java:1699)
        at com.sun.org.apache.xerces.internal.impl.XMLEntityScanner.scanData(java.xml@22.0.2/XMLEntityScanner.java:1256)
        at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanCDATASection(java.xml@22.0.2/XMLDocumentFragmentScannerImpl.java:1637)
        at com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(java.xml@22.0.2/XMLDocumentFragmentScannerImpl.java:2957)
        at com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(java.xml@22.0.2/XMLDocumentScannerImpl.java:635)
        at com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(java.xml@22.0.2/XMLNSDocumentScannerImpl.java:113)
        at com.sun.org.apache.xerces.internal.impl.XMLStreamReaderImpl.next(java.xml@22.0.2/XMLStreamReaderImpl.java:601)
        at com.apptasticsoftware.rssreader.AbstractRssReader$RssItemIterator.next(AbstractRssReader.java:563)
        at com.apptasticsoftware.rssreader.AbstractRssReader$RssItemIterator.peekNext(AbstractRssReader.java:537)
        at com.apptasticsoftware.rssreader.AbstractRssReader$RssItemIterator.hasNext(AbstractRssReader.java:547)
        at java.util.Iterator.forEachRemaining(java.base@22.0.2/Iterator.java:132)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(java.base@22.0.2/Spliterators.java:1939)
        at java.util.stream.AbstractPipeline.copyInto(java.base@22.0.2/AbstractPipeline.java:556)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(java.base@22.0.2/AbstractPipeline.java:546)
        at java.util.stream.AbstractPipeline.evaluate(java.base@22.0.2/AbstractPipeline.java:622)
        at java.util.stream.AbstractPipeline.evaluateToArrayNode(java.base@22.0.2/AbstractPipeline.java:291)
        at java.util.stream.ReferencePipeline.toArray(java.base@22.0.2/ReferencePipeline.java:631)
        at java.util.stream.ReferencePipeline.toArray(java.base@22.0.2/ReferencePipeline.java:637)
        at java.util.stream.ReferencePipeline.toList(java.base@22.0.2/ReferencePipeline.java:642)
        at io.github.arlol.RssToMailApplication.process(RssToMailApplication.java:84)
```

**Comment by w3stling** [comment]
Seems like it is not possible to do that with `java.net.http.HttpClient`.
There exist a ticket to add timeout for reading body but it has not been implemented.
https://bugs.openjdk.org/browse/JDK-8258397

You can use a different http client:
```java
var url = new URL("https://lwn.net/headlines/rss");
var urlConnection = url.openConnection();
urlConnection.setReadTimeout(TimeUnit.SECONDS.toMillis(25));
var items = new RssReader().read(urlConnection.getInputStream()).collect(Collectors.toList());
```

**Comment by w3stling** [comment]
Added read timeout in PR #169. Included in [v3.8.0 release](https://github.com/w3stling/rssreader/releases/tag/v3.8.0)

If using the RssReader as before a default read timeout of 25 seconds is used.
```java
var items = new RssReader().read("https://lwn.net/headlines/rss")
                           .collect(Collectors.toList());
```

To change the default read timeout to 10 seconds:
```java
var items = new RssReader().read("https://lwn.net/headlines/rss")
                           .setReadTimeout(Duration.ofSeconds(10))
                           .collect(Collectors.toList());
```

If an HTTP client that supports read timeout is used and an InputStream is passed to RssReader, the built-in read timeout can be disabled by setting the timeout to Duration.ZERO or set to a higher value than set for the HTTP client.

Please try this and let me know if this solves your problem.

**Comment by ArloL** [comment]
Thanks so much. I went ahead and [started using apache httpclient5](https://github.com/ArloL/rss-to-mail/commit/46e1513c3b30c91c1679b000b361d2f38e0c7714) as per your suggestion. I will use both and open another issue if I have any problems.

Thanks!
